Setting Admin - Add mixin to autogenerate a settings form

totten opened the pull request:
Overview
----------------------------------------

@mattwire @eileenmcnaughton @seamuslee001 - I was reading https://docs.civicrm.org/dev/en/latest/framework/setting/extension/, and it struck me that many of these steps could be done automatically...

Before
----------------------------------------

Suppose you are creating extension `foobar`. It needs some settings and a setting page. Process:

1. Create the extension (`civix generate:module foobar`)
2. Create the settings (`nano settings/foobar.setting.php`)
3. Implement `hook_civicrm_permission` and declare `administer foobar`
4. Create `xml/Menu/foobar.xml` and define route `civicrm/admin/setting/foobar`. Enable mixin `menu-xml@1`.
5. Update each setting to declare `settings_page => ['foobar' => ['weight' => 123]]` (*with differing values of 123*)

(*And, of course, clear caches.*)
 
After
----------------------------------------

The last several steps are consolidated into one step. Process:

1. Create the extension (`civix generate:module foobar`)
2. Create the settings (`nano settings/foobar.setting.php`)
3. In `info.xml`, add `<mixin>setting-admin@1</mixin>`.

(*And, of course, clear caches.*)

Comments
----------------------------------------

Example extension: https://lab.civicrm.org/-/snippets/95

This mixin should be amenable to backporting. So we could add it to `civix` and use it for extensions that run on older versions of `civicrm-core`. I don't know how far back compatibility would go... but probably several years. For DX, the process would be like:

```
$ civix mixin --enable=setting-admin@1,setting-php@1
Updating info.xml
Writing mixin/setting-admin@1.0.0.mixin.php
Writing mixin/setting-php@1.0.0.mixin.php
```

If the concept seems good, then we should add some tests to the mixin.

totten committed:
setting-admin@1 - Initial implementation of mixin

eileenmcnaughton commented:
@totten yep - looks good!

mattwire commented:
So the path is always the extension key? Sounds reasonable just not clear from description.

I'm so used to "wiring" up extension settings pages I hardly think about it but it's a lot of steps and this looks really nice + simplifies extension + makes a transition to a non-quickform settings page easy in the future (ie. just replace the mixin).

totten commented:
> So the path is always the extension key? Sounds reasonable just not clear from description.

Yeah, I had done the path with the `key` (long name). But everything else used the `file` (short name). And I think the short name is probably more typical of existing/real-world settings-pages. So I'll switch it to match.

totten committed:
mixer - Fix warning on php81

totten committed:
setting-admin@1 - Generate nav-link

totten committed:
authx - Switch to setting-admin@1

totten committed:
recaptcha - Switch to setting-admin@1

totten committed:
setting-admin@1 - Add E2E tests

totten commented:
OK, multiple updates:

* Standardized on the short-name (eg `myext` instead of`org.example.myext`) for paths/permissions/etc.
* Added E2E tests
* Added nav-link
* Converted a couple of core-exts (`authx` and `recaptcha`) to use it
* Did some light `r-run` in other languages. (*The page-titles -- "Authentication" and "reCAPTCHA Settings" -- don't appear to be in my `l10n/fr_CA` data. But if I hack `info.xml` and use some well-known strings, then I can see that the `ts()` mechanism is being applied - same as it was before the patch.*)
* Resolved MC from recent PR

eileenmcnaughton commented on mixin/setting-admin@1/mixin.php:
```diff
@@ -0,0 +1,211 @@
+   * @return string|null
+   *   Return NULL by default. If the walk was short-circuited, then return that value.
+   */
+  public static function walk(&$items, callable $callback) {
```
is this like run but slower?

totten replied:
rofl. Yeah, basically. 

FWIW, for an example of this wording, see https://www.php.net/array_walk or https://www.php.net/array_walk_recursive. It's the same basic idea, except that it's tuned to the properties used by nav-menu (e.g. `$item`, `$item['child']`, `$item['attribute']`).

highfalutin commented:
Is there a docs MR for this?

The relevant docs page looks to be [Creating a extension setting and admin form](https://docs.civicrm.org/dev/en/latest/framework/setting/extension/) and text could be borrowed from the [code doc](https://github.com/civicrm/civicrm-core/pull/27569/files#diff-d563a8702160cb93a618c65851be875e4b7f6123cb6d16fa043fe28234e850abR4-R23)
